### PR TITLE
[DISCO-2649] Suggest: Amo keyword logic to match prefix and suffix

### PR DESCRIPTION
**Pull request by tiftran**
### Pull Request checklist ###

- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [ ] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

File: components/suggest/src/db.rs [hunk 1]
```diff
@@ -128,7 +128,9 @@ impl<'a> SuggestDao<'a> {
         let (keyword_prefix, keyword_suffix) = split_keyword(&query.keyword);
         let suggestions_limit = query.limit.unwrap_or(-1);
 
-        let (mut statement, params) = if query.providers.contains(&SuggestionProvider::Pocket) {
+        let (mut statement, params) = if query.providers.contains(&SuggestionProvider::Pocket)
+            || query.providers.contains(&SuggestionProvider::Amo)
```
Comment (ncloudioj) [inline components/suggest/src/db.rs:132]: Nit: you can also do this with the benifit that it will scan the provider array once.
```suggestion
        let (mut statement, params) = if query
            .providers
            .iter()
            .any(|p| matches!(p, SuggestionProvider::Pocket | SuggestionProvider::Amo))

```

File: components/suggest/src/db.rs [hunk 2]
```diff
@@ -141,6 +143,11 @@ impl<'a> SuggestDao<'a> {
                      FROM suggestions s
                      JOIN pocket_keywords k ON k.suggestion_id = s.id
                      WHERE k.keyword_prefix = :keyword_prefix
+                     UNION ALL
```
Comment (ncloudioj) [inline components/suggest/src/db.rs:146]: This will query both amo and pocket keyword tables despite that the consumer might need one of them. I think we have two options here:
- Merge `amo_keywords` and `pocket_keywords` as they're almost identical with the same matching logic. We can split them up if they diverge in the future
- Split this SQL into multiple so that it will query suggestions per provider, which means it'd issue multiple calls against the database. The upside is that we don't need to maintain this SQL for different keyword tables

File: components/suggest/src/db.rs (hunk 1 above)
Comment (tiftran) [reply components/suggest/src/db.rs:132]: oooh fancy

File: components/suggest/src/db.rs (hunk 2 above)
Comment (tiftran) [reply components/suggest/src/db.rs:146]: makes sense, let me try to fiddle with both options

Comment (tiftran) [reply components/suggest/src/db.rs]: okie doke, so i think i was able to implement the first option, currently trying the second

File: components/suggest/src/db.rs [hunk 3]
```diff
@@ -242,6 +247,7 @@ impl<'a> SuggestDao<'a> {
                                 ":suggestion_id": suggestion_id
                             },
                             |row| {
+                                if full_suffix.contains(keyword_suffix){
```
Comment (linabutler) [inline components/suggest/src/db.rs:250]:
```suggestion
                                if full_suffix.contains(keyword_suffix) {
```

File: components/suggest/src/schema.rs [hunk 4]
```diff
@@ -21,10 +21,10 @@ pub const SQL: &str = "
         PRIMARY KEY (keyword, suggestion_id)
     ) WITHOUT ROWID;
 
-    CREATE TABLE pocket_keywords( 
+    CREATE TABLE prefix_keywords(
         keyword_prefix TEXT NOT NULL,
         keyword_suffix TEXT NOT NULL DEFAULT '',
-        confidence INTEGER NOT NULL,
+        confidence INTEGER NOT NULL DEFAULT '',
```
Comment (linabutler) [inline components/suggest/src/schema.rs:27]:
```suggestion
        confidence INTEGER NOT NULL DEFAULT 0,
```

`''` is technically legal—SQLite's type system is [relaxed by default](https://www.sqlite.org/datatype3.html)—but let's use a number if the column is declared as a number.

File: components/suggest/src/store.rs [hunk 5]
```diff
@@ -1945,14 +1945,34 @@ mod tests {
                 "#]],
             ),
             (
-                "keyword = `soft`; AMP, Wikipedia, and AMO",
+                "keyword = `masking e`; AMO only",
```
Comment (linabutler) [inline components/suggest/src/store.rs:1948]: Just for completeness, could you please add another test for one of the single-word keywords, like `relay` or `spam`?

**Review by linabutler — APPROVED** [review]
r+wc, thanks! This looks like it'll be a very easy uplift 😊

File: components/suggest/src/db.rs [hunk 6]
```diff
@@ -242,6 +247,7 @@ impl<'a> SuggestDao<'a> {
                                 ":suggestion_id": suggestion_id
                             },
                             |row| {
+                                if full_suffix.contains(keyword_suffix) {
```
Comment (linabutler) [inline components/suggest/src/db.rs:250]: Sorry, I just noticed this: `.contains()` will match anywhere in the string, so `masking mail`, `masking i`, `masking il`, `masking l`, etc. will all match `masking email`.

Is that intentional? Or do we want this to be a prefix match, so only `masking e`, `masking em`, `masking ema`, etc. should match?

(Either way, let's add another test for that case, too, please—either a positive one if it's intentional, or a negative one that verifies we don't return anything. Thanks!)

Comment (linabutler) [reply components/suggest/src/db.rs:250]: Side note: I also really like how you factored out `full_suffix` into a variable, and wondering if we can give the logic in the Pocket match arm the same treatment, while you're here?

```rs
let confidence = row.get("confidence")?;
let full_suffix = row.get::<_, String>("keyword_suffix")?;
let suffixes_match = match confidence {
    KeywordConfidence::Low => full_suffix.starts_with(keyword_suffix),
    KeywordConfidence::High => full_suffix == keyword_suffix,
};
```

**Review by linabutler — APPROVED** [review]
Thank you! 🙌🏼
